param.go

```diff
@@ -0,0 +1,72 @@
+package dig
+
+import (
+	"errors"
+	"reflect"
+)
+
+// Param represents an argument to invoked function.
+type Param struct {
```

Comment:
Interesting. But now caller will have to infer and pass in `reflect.Type`, `reflect.Value` as well. How about `Param` just be a variadic interface and we do the heavy loading of finding Type and Value?